In [23]:
import pandas as pd

historical_transactions = pd.read_csv("data/Historical-transaction-data.csv")
store_info = pd.read_csv("data/Store-info.csv")

# Historical-transaction-data.csv Analysis

In [24]:
historical_transactions.head()

item_description          transaction_date  invoice_id customer_id   
0     ORANGE BARLEY 1.5L  2021-12-11T00:00:00.000Z       147.0        BGXA  \
1       GINGER BEER 1.5L  2021-10-17T00:00:00.000Z       371.0        IA25   
2        TONIC PET 500ML  2021-12-13T00:00:00.000Z       484.0        VN7V   
3          CREAM SODA 1L  2021-12-13T00:00:00.000Z       484.0        VN7V   
4  STRAWBERRY MILK 180ML  2021-10-23T00:00:00.000Z      1310.0        7S00   

   shop_id  item_price  quantity_sold  
0  SHOP008         220              2  
1  SHOP112         220              2  
2  SHOP008         160              2  
3  SHOP008         150              2  
4  SHOP112         210              5

In [25]:
historical_transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473974 entries, 0 to 473973
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   item_description  438046 non-null  object 
 1   transaction_date  473974 non-null  object 
 2   invoice_id        467654 non-null  float64
 3   customer_id       473974 non-null  object 
 4   shop_id           473974 non-null  object 
 5   item_price        473974 non-null  int64  
 6   quantity_sold     473974 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 25.3+ MB


In [26]:
historical_transactions.describe()

invoice_id     item_price  quantity_sold
count  4.676540e+05  473974.000000  473974.000000
mean   1.996684e+06     206.689734       1.925506
std    1.344594e+06     166.764732       1.634535
min    1.700000e+01      35.000000      -1.000000
25%    1.032114e+06     100.000000       1.000000
50%    2.032996e+06     200.000000       2.000000
75%    3.032568e+06     220.000000       2.000000
max    8.331754e+06   17400.000000     101.000000

In [ ]:
# TODO: remove outliers

In [27]:
print("\n* Historical-transaction-data.csv SUMMARY *\n")
# Column names
print(f"{', '.join(historical_transactions.columns.tolist())} are the column names")
# No of rows
print(f"{len(historical_transactions)} total rows")

# Unique shop ids
shop_ids = historical_transactions["shop_id"].unique()
print(f"{len(shop_ids)} unique shop_id")

# Unique items
item_ids = historical_transactions["item_description"].unique()
print(f"{len(item_ids)} unique item_description")

# Unique transaction date
transaction_dates = historical_transactions["transaction_date"].unique()
print(f"{len(transaction_dates)} unique transaction_date")

# Unique item price
item_prices = historical_transactions["item_price"].unique()
print(f"{len(item_prices)} unique item_price")

# Unique quantity sold
quantities_sold = historical_transactions["quantity_sold"].unique()
print(f"{len(quantities_sold)} unique quantity_sold")

# Column names containing null or missing values
print(f"Only {', '.join(historical_transactions.columns[historical_transactions.isnull().any()].tolist())} columns contains empty values")

# No of rows in item_description containing null or missing values
print(f"{len(historical_transactions[historical_transactions['item_description'].isnull()])} rows contain empty values in item_description")

# No of rows in invoice_id containing null or missing values
print(f"{len(historical_transactions[historical_transactions['invoice_id'].isnull()])} rows contain empty values in invoice_id")

# No of rows in quantity_sold with -1 value
print(f"{len(historical_transactions[historical_transactions['quantity_sold'] <= 0])} rows contain 0 or less value in quantity_sold")

# No of rows with null or missing values
print(f"{len(historical_transactions[historical_transactions.isnull().any(axis=1)])} total rows with empty values")


* Historical-transaction-data.csv SUMMARY *

item_description, transaction_date, invoice_id, customer_id, shop_id, item_price, quantity_sold are the column names
473974 total rows
124 unique shop_id
38 unique item_description
62 unique transaction_date
209 unique item_price
69 unique quantity_sold
Only item_description, invoice_id columns contains empty values
35928 rows contain empty values in item_description
6320 rows contain empty values in invoice_id
8351 rows contain 0 or less value in quantity_sold
41865 total rows with empty values


## Historical-transaction-data.csv SUMMARY

### Handling Missing Values:
- For the missing item_description values, since there are 38 unique item descriptions, you may choose to fill the missing values with the mode (the most frequent item description) or drop the rows with missing values. Considering there are 35,928 rows with missing values, filling with the mode might be a reasonable choice.
- For the missing invoice_id values, you can either fill them using the median or drop the rows with missing values. Since there are 6,320 rows with missing values, filling with the median might be a reasonable choice.
- For the rows with 0 or less value in quantity_sold, you should consider dropping these rows, as zero or negative quantity does not make sense in this context.

### Feature Engineering:
- Extract relevant features from the transaction_date column, such as month, day, or day of the week.
- Create new features that might be useful for the analysis, such as total sales per store, average sales per customer, etc.

### Data Transformation:
- Convert categorical variables into numerical representations, like one-hot encoding or label encoding, if required by your chosen model.
- Normalize or standardize numerical features if needed.

# Store-info.csv Analysis

In [28]:
store_info.head()

shop_id  shop_area_sq_ft shop_profile
0  SHOP047              528     Moderate
1  SHOP009              676         High
2  SHOP083              676          Low
3  SHOP117              676          Low
4  SHOP042              676          Low

In [29]:
store_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   shop_id          124 non-null    object
 1   shop_area_sq_ft  124 non-null    int64 
 2   shop_profile     100 non-null    object
dtypes: int64(1), object(2)
memory usage: 3.0+ KB


In [30]:
store_info.describe()

shop_area_sq_ft
count       124.000000
mean        619.991935
std         126.770165
min         298.000000
25%         553.250000
50%         617.000000
75%         676.000000
max        1077.000000

In [31]:
print("\n* Store-info.csv SUMMARY *\n")
# Column names
print(f"{', '.join(store_info.columns.tolist())} are the column names")
# No of rows
print(f"{len(store_info)} total rows")

# Unique shop ids
shop_ids_store = store_info["shop_id"].unique()
print(f"{len(shop_ids_store)} unique shop_id")

# Unique shop area
shop_area = store_info["shop_area_sq_ft"].unique()
print(f"{len(shop_area)} unique shop_area_sq_ft")

# Unique shop profiles
shop_profiles = store_info["shop_profile"].unique()
print(f"{', '.join([ item for item in shop_profiles.tolist() if type(item) == str ])} shop_profile values")

# Column names containing null or missing values
print(f"Only {', '.join(store_info.columns[store_info.isnull().any()].tolist())} columns contains empty values")

# No of rows with null or missing values
print(f"{len(store_info[store_info.isnull().any(axis=1)])} total rows with empty values")


* Store-info.csv SUMMARY *

shop_id, shop_area_sq_ft, shop_profile are the column names
124 total rows
124 unique shop_id
66 unique shop_area_sq_ft
Moderate, High, Low shop_profile values
Only shop_profile columns contains empty values
24 total rows with empty values


## Store-info.csv SUMMARY

### Handling Missing Values:
- For the missing shop_profile values, you do not need to fill them in this dataset, as these are the shops you need to predict the shop profiles for in the Testing-data.csv. You can simply leave them as they are.

### Feature Engineering:
- You may want to create new features by combining information from the Historical-transaction-data.csv dataset. For example, you can calculate the total sales, average sales per customer, or number of unique customers per store, and add them to the Store-info.csv dataset. This will provide additional information for your model to make better predictions.

### Data Transformation:
- If needed, normalize or standardize numerical features like shop_area_sq_ft.

After cleaning and preprocessing the data for both datasets, you can join them together on the shop_id column, and then proceed with model selection, training, and evaluation.